lib/auth/winwebauthn/api.go

```diff
@@ -0,0 +1,221 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package winwebauthn is wrapper around Windows webauthn API.
+// It loads system webauthn.dll and uses it's method.
+// It supports API versions 1-4.
+// API definition: https://github.com/microsoft/webauthn/blob/master/webauthn.h
+// As Windows Webauthn device can be used both Windows Hello and FIDO devices.
+package winwebauthn
+
+import (
+	"context"
+	"io"
+
+	"github.com/duo-labs/webauthn/protocol"
+	"github.com/duo-labs/webauthn/protocol/webauthncose"
+	"github.com/gravitational/teleport/api/client/proto"
+	wanlib "github.com/gravitational/teleport/lib/auth/webauthn"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// LoginOpts groups non-mandatory options for Login.
+type LoginOpts struct {
+	// AuthenticatorAttachment specifies the desired authenticator attachment.
+	AuthenticatorAttachment AuthenticatorAttachment
+}
+
+type AuthenticatorAttachment int
+
+const (
+	AttachmentAuto AuthenticatorAttachment = iota
+	AttachmentCrossPlatform
+	AttachmentPlatform
+)
+
+// nativeWebauthn represents the native windows webauthn interface.
+// Implementors must provide a global variable called `native`.
+type nativeWebauthn interface {
+	CheckSupport() CheckSupportResult
+	GetAssertion(origin string, in protocol.PublicKeyCredentialRequestOptions, loginOpts *LoginOpts) (*wanlib.CredentialAssertionResponse, error)
+	MakeCredential(origin string, in protocol.PublicKeyCredentialCreationOptions) (*wanlib.CredentialCreationResponse, error)
+}
+
+// Login implements Login for Windows Webauthn API.
+// It returns an MFAAuthenticateResponse.
+// Most callers should call Login directly, as it is correctly guarded by
+// IsAvailable.
+func Login(ctx context.Context, origin string, assertion *wanlib.CredentialAssertion, opts *LoginOpts) (*proto.MFAAuthenticateResponse, string, error) {
+	switch {
+	case origin == "":
+		return nil, "", trace.BadParameter("origin required")
+	case assertion == nil:
+		return nil, "", trace.BadParameter("assertion required")
+	case len(assertion.Response.Challenge) == 0:
+		return nil, "", trace.BadParameter("assertion challenge required")
+	case assertion.Response.RelyingPartyID == "":
+		return nil, "", trace.BadParameter("assertion relying party ID required")
+	}
+	resp, err := native.GetAssertion(origin, assertion.Response, opts)
+	if err != nil {
+		// TODO(tobiaszheller): proper error
+		return nil, "", err
```

Comment:
```suggestion
		return nil, "", trace.Wrap(err)
```

Same where applicable.

Reply:
Bump so it won't be forgotten.